lib/db/db.rb

```diff
@@ -9,26 +9,63 @@ def size
     all.size
   end
 
+  def all
+    reader.nil? ? [] : reader.run
+  end
+
+  def at(index)
+    empty || invalid(index) ? {} : all[index]
+  end
+
   def add(contact)
-    contacts = all
-    contacts << contact
+    contacts = all << format(contact)
     writer.run(contacts) if !writer.nil?
   end
 
-  def all
-    reader.nil? ? [] : reader.run
+  def update(contact)
+    contacts = updated_with(format(contact))
+    writer.run(contacts) if !writer.nil?
   end
 
-  def at(id)
-    all.empty? ? {} : all[id]
+  def search(term)
+    all.select { |contact| contact.any? {|key, val| val.to_s.include?(term)} }
   end
 
-  def search(term)
-    all.select{ |contact| contact.any? {|key, val| val.include?(term)} }
+  def index_of_id(id)
+    all.index { |contact| contact["id"] == id }
   end
 
   private
 
   attr_reader :reader, :writer
 
+  def empty
+    all.empty?
+  end
+
+  def invalid(index)
+    index.nil? ? true : all[index].nil?
+  end
+
+  def format(contact)
+    contact.has_key?("id") ? contact : append_id(contact, id_of_new)
```

Comment:
Ah I see. Yeah, good point, the creator shouldn't know about an ID :+1: 
Yet, the term `format` has a different connotation. I think of format, as in formatting something. Not adding an ID. Would something like `add_id` or `ensure_id` make it more obvious what's happening?